src/6_branch.js

```diff
@@ -67,16 +67,23 @@ var wrap = function(parameters, func, init) {
 			 * @type {function(?Error,?): undefined}
 			 */
 			var done = function(err, data) {
-				if (err && parameters === callback_params.NO_CALLBACK) {
+				try {
+					if (err && parameters === callback_params.NO_CALLBACK) {
+						throw err;
+					}
+					else if (parameters === callback_params.CALLBACK_ERR) {
+						callback(err);
+					}
+					else if (parameters === callback_params.CALLBACK_ERR_DATA) {
+						callback(err, data);
+					}
+				} catch (err) {
+					// the callback functions may or may not throw an error...
 					throw err;
```

Comment:
The key is that `next` needs to be called in all cases, but we still also need to bubble errors up to the call stack.  Without that `catch` clause does it feel better to you?